python/cudf/cudf/core/column/categorical.py

```diff
@@ -1062,51 +1062,36 @@ def notnull(self) -> ColumnBase:
-        method: Optional[str] = None,
-    ) -> Self:
-        """
-        Fill null values with *fill_value*
-        """
-        if fill_value is not None:
-            fill_is_scalar = np.isscalar(fill_value)
-
-            if fill_is_scalar:
-                if fill_value == _DEFAULT_CATEGORICAL_VALUE:
-                    fill_value = self.codes.dtype.type(fill_value)
-                else:
-                    try:
-                        fill_value = self._encode(fill_value)
-                        fill_value = self.codes.dtype.type(fill_value)
-                    except ValueError as err:
-                        err_msg = "fill value must be in categories"
-                        raise ValueError(err_msg) from err
-            else:
-                fill_value = column.as_column(fill_value, nan_as_null=False)
-                if isinstance(fill_value, CategoricalColumn):
-                    if self.dtype != fill_value.dtype:
-                        raise TypeError(
-                            "Cannot set a Categorical with another, "
-                            "without identical categories"
-                        )
-                # TODO: only required if fill_value has a subset of the
-                # categories:
-                fill_value = fill_value._set_categories(
-                    self.categories,
-                    is_unique=True,
-                )
-                fill_value = column.as_column(fill_value.codes).astype(
-                    self.codes.dtype
+    def _validate_fillna_value(
+        self, fill_value: ScalarLike | ColumnLike
+    ) -> cudf.Scalar | ColumnBase:
+        """Align fill_value for .fillna based on column type."""
+        if cudf.api.types.is_scalar(fill_value):
+            if fill_value != _DEFAULT_CATEGORICAL_VALUE:
+                try:
+                    fill_value = self._encode(fill_value)
+                except ValueError as err:
+                    raise ValueError(
+                        f"{fill_value=} must be in categories"
+                    ) from err
+            return cudf.Scalar(fill_value, dtype=self.codes.dtype)
+        else:
+            fill_value = column.as_column(fill_value, nan_as_null=False)
+            if (
+                isinstance(fill_value.dtype, CategoricalDtype)
+                and self.dtype != fill_value.dtype
+            ):
+                raise TypeError(
+                    "Cannot set a Categorical with another, "
+                    "without identical categories"
                 )
-
-        # Validation of `fill_value` will have to be performed
-        # before returning self.
-        if not self.nullable:
-            return self
-
-        return super().fillna(fill_value, method=method)
+            # TODO: only required if fill_value has a subset of the
+            # categories:
+            fill_value = fill_value._set_categories(
+                self.categories,
+                is_unique=True,
+            )
```

Comment:
I think this is only the case if `fill_value.dtype` is _not_ categorical after the `as_column` call, right?
So perhaps:
```python
fill_value = column.as_column(...)
if isinstance(fill_value.dtype, CategoricalDtype):
    if self.dtype != fill_value.dtype: raise
else:
    fill_value = fill_value._set_categories(self.categories, is_unique=True)
```

But, question: does `set_categories` work on any column, not just a categoricalcolumn, and should `is_unique` be dependent on `self.categories`?

Reply:
> does set_categories work on any column

Looks like this is only defined on a `CategoricalColumn` so `fill_value` must be one here

> and should is_unique be dependent on self.categories?

IIUC, `is_unique` pretty much means "call `unique` on the input if it's not unique" so I _think_ the `is_unique=True` might not be needed here